controllers/deactivation/deactivation_controller.go

```diff
@@ -193,9 +240,24 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 		return reconcile.Result{}, nil
 	}
 
+	// We calculate the actual deactivation due time based on when the deactivating condition was set.  This may end up
+	// being significantly later than the scheduled deactivation time set in UserSignup.Status.ScheduledDeactivationTime, as
+	// in some rare circumstances the deactivating notification/status may fail to be set due to cluster downtime or
+	// other reasons.  Because of this, the scheduled deactivation time that is set in the UserSignup.Status should be
+	// treated as informational only.
 	deactivationDueTime := deactivatingCondition.LastTransitionTime.Time.Add(time.Duration(deactivatingNotificationDays*24) * time.Hour)
```

Comment:
We actually should use this `deactivationDueTime` above before returning `reconcile.Result{}, nil` in line 240 as the scheduled deactivation time.
So if the actual deactivation is postponed because the deactivating notification condition is not `true` yet (so the notification has not been created yet) then we need to update the scheduled time too.

Reply:
I'm a bit confused by this one - if the deactivating condition is not found, or it is not in an expected state then it cannot be used as a basis for calculating the scheduled deactivation time.  In this case, the only other valid way of calculating it is based on the provisioned time, which we already do above.  Or am I missing something here?

Reply:
This is my thinking.
Let's say it's time to start the deactivation but the the deactivating condition is not set or is not in an expected state. Meaning we have to delay the deactivation. Something like this for example:
- Provisioning time: - 30 days (30 days ago).
- The original scheduled deactivation time: now (meaning 0 days left).
- The deactivating condition is still pending (there is an unexpected delay in the notification mechanism for example).
With the current code we will still show that there is no time left before the deactivation happens. While in the reality we won't start the deactivation for *at least* three more days. It maybe be longer than 3 days but it's impossible to know. So our best guess would be to use the same `now`  + `deactivatingNotificationDays` (which is currently 3 days).